### PR TITLE
Tag Processor: Add ability to stop at tag closers, if requested

### PR DESCRIPTION
**Pull request by dmsnell**
## Notes

 - @adamziel for now I've added new tests to the existing and incorrect test suite. once we land #45762 or #44593 I'll update this PR to reflect the unified/merged test files.

## What?

Add an option to the `WP_HTML_Tag_Processor` query to indicate that the processor should stop and visit tag closers while walking through the input document.

## Why?

`WP_HTML_Tag_Processor` introduces the ability to walk through an HTML document and modify attributes on tag openers.

At times it could be useful to stop also at tag closers in order to perform augmented inspection and querying of a document.

In this patch we're opening up a mode that allows just that; if querying with the `[ "tag_closers" => "visit" ]` query parrameter then in addition to stopping at tag openers, this will allow stopping at all tag tokens regardless of open or close status.

## How?

Introduces internal mechanics to detect and stop at tag closers.
Exposes new query option `"tag_closers" => "visit" | "skip"` to pass into `next_tag()` which indicates whether or not to stop at tag closers.
Exposes new `is_tag_closer()` method to determine if current tag is an opener or closer (from calling code).

## Testing Instructions

Review the new unit tests to confirm behavior.

### CODE REVIEW

File: lib/experimental/html/class-wp-html-tag-processor.php [hunk 1]
```diff
@@ -408,7 +439,16 @@ public function next_tag( $query = null ) {
 				return false;
 			}
 
-			$this->parse_tag_opener_attributes();
+			while ( $this->parse_next_attribute() ) {
```
Comment (adamziel) [inline lib/experimental/html/class-wp-html-tag-processor.php:446]: I like the inlining

File: lib/experimental/html/class-wp-html-tag-processor.php [hunk 2]
```diff
@@ -1117,6 +1148,25 @@ public function get_tag() {
...
+	 *
+	 * Example:
+	 * <code>
+	 *     $p = new WP_HTML_Tag_Processor( '<div></div>' );
+	 *     $p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] );
+	 *     $p->is_tag_closer() === false;
+	 *
+	 *     $p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] );
+	 *     $p->iss_tag_closer() === true;
+	 * </code>
```
Comment (adamziel) [inline lib/experimental/html/class-wp-html-tag-processor.php]: Nice example! A minor typo:

```suggestion
	 * Example:
	 * <code>
	 *     $p = new WP_HTML_Tag_Processor( '<div></div>' );
	 *     $p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] );
	 *     $p->is_tag_closer() === false;
	 *
	 *     $p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] );
	 *     $p->is_tag_closer() === true;
	 * </code>
```

File: phpunit/html/WP_HTML_Tag_Processor_Test.php [hunk 3]
```diff
@@ -217,6 +217,22 @@ public function test_next_tag_should_return_false_for_a_non_existing_tag() {
...
+	public function test_next_tag_should_stop_on_closers_only_when_requested() {
+		$p = new WP_HTML_Tag_Processor( '<div><img /></div>' );
+		$this->assertTrue( $p->next_tag( [ 'tag_name' => 'div' ] ), 'Did not find desired tag opener' );
+		$this->assertFalse( $p->next_tag( [ 'tag_name' => 'div' ] ), 'Visited an unwanted tag, a tag closer' );
+
+		$p = new WP_HTML_Tag_Processor( '<div><img /></div>' );
+		$p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] );
+		$this->assertFalse( $p->is_tag_closer(), 'Indicated a tag opener is a tag closer' );
+		$this->assertTrue( $p->next_tag( [ 'tag_name' => 'div', 'tag_closers' => 'visit' ] ), 'Did not stop at desired tag closer' );
+		$this->assertTrue( $p->is_tag_closer(), 'Indicated a tag closer is a tag opener' );
```
Comment (adamziel) [inline phpunit/html/WP_HTML_Tag_Processor_Test.php]: I'd like to see some tests to confirm the behavior of the other methods in the tag closer state, e.g. `update_attribute()`, `__toString()`, etc

**Comment by adamziel** [comment]
This looks good @dmsnell! There are some conflicts and I left a few comments, but the direction is great.

File: phpunit/html/WP_HTML_Tag_Processor_Test.php (hunk 3 above)
Comment (dmsnell) [reply phpunit/html/WP_HTML_Tag_Processor_Test.php]: added a test to confirm that ops are inert inside closed tags

**Review by adamziel — APPROVED** [review]
Thank you for the updates @dmsnell! The linter fails, but the tests are passing. This is a solid PR that makes a great foundation for the upcoming unsafe mode. 🚢

**Comment by dmsnell** [comment]
Fixed the lint issues

**Pull request merged**